hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/CompoundNaturalIdMapping.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.metamodel.mapping.internal;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+import org.hibernate.NotYetImplementedFor6Exception;
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.loader.NaturalIdPostLoadListener;
+import org.hibernate.loader.NaturalIdPreLoadListener;
+import org.hibernate.loader.ast.internal.CompoundNaturalIdLoader;
+import org.hibernate.loader.ast.internal.MultiNaturalIdLoaderStandard;
+import org.hibernate.loader.ast.spi.MultiNaturalIdLoader;
+import org.hibernate.loader.ast.spi.NaturalIdLoader;
+import org.hibernate.metamodel.mapping.ColumnConsumer;
+import org.hibernate.metamodel.mapping.EntityMappingType;
+import org.hibernate.metamodel.mapping.JdbcMapping;
+import org.hibernate.metamodel.mapping.MappingType;
+import org.hibernate.metamodel.mapping.SingularAttributeMapping;
+import org.hibernate.query.NavigablePath;
+import org.hibernate.sql.ast.Clause;
+import org.hibernate.sql.ast.spi.SqlSelection;
+import org.hibernate.sql.ast.tree.from.TableGroup;
+import org.hibernate.sql.results.graph.DomainResult;
+import org.hibernate.sql.results.graph.DomainResultCreationState;
+import org.hibernate.type.descriptor.java.JavaTypeDescriptor;
+import org.hibernate.type.spi.TypeConfiguration;
+
+/**
+ * Multi-attribute NaturalIdMapping implementation
+ */
+public class CompoundNaturalIdMapping extends AbstractNaturalIdMapping implements MappingType {
+
+	// todo (6.0) : create a composite MappingType for this descriptor's Object[]?
+
+	private final List<SingularAttributeMapping> attributes;
+	private final List<JdbcMapping> jdbcMappings;
+
+	private final NaturalIdLoader<?> loader;
+	private final MultiNaturalIdLoader<?> multiLoader;
+
+	public CompoundNaturalIdMapping(
+			EntityMappingType declaringType,
+			List<SingularAttributeMapping> attributes,
+			String cacheRegionName,
+			MappingModelCreationProcess creationProcess) {
+		super( declaringType, cacheRegionName );
+		this.attributes = attributes;
+
+		final List<JdbcMapping> jdbcMappings = new ArrayList<>();
```

Comment:
Could we make an educated guess about the list size based on attributes? Seems most of the time it is predicatable.